### PR TITLE
Fix Base64 duplicate padding

### PR DESCRIPTION
**Pull request by JunielKatarn**
## Description
Ensure there is no `=` padding for encoded Base64 values with size multiple of 4.

### Type of Change
- Bug fix (non-breaking change which fixes an issue)

### Why
Current Boost-based Base64 encoding used in the `FileReader` module will pad four `=` characters for encoded strings with length multiple of 4.
This padding is redundant and constitutes incorrect output.

### What
Centralize the Boost-based Base64 functions into:
- `Microsoft::React::Utilities::EncodeBase64`
- `Microsoft::React::Utilities::DecodeBase64`

Use functions above where applicable (Base64 encoding/decoding that does not have a hard dependency on `IBuffer`).

MSBuild project updates:
- React.Windows.Common
  - Added dependency on `boost` and `cppwinrt`.

## Testing
Defined tests:
- `Microsoft::React::Test::UtilsTest::EncodeStdStringToBase64Succeeds`
- `Microsoft::React::Test::UtilsTest::DecodeStdStringFromBase64Succeeds`

### CODE REVIEW

File: vnext/Common/Utilities.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  auto bytes = av_t(text.data(), static_cast<av_t::size_type>(text.size()));
+
+  using namespace boost::archive::iterators;
+  typedef base64_from_binary<transform_width<const char *, 6, 8>> encode_base64;
+  std::ostringstream oss;
+  std::copy(encode_base64(bytes.cbegin()), encode_base64(bytes.cend()), ostream_iterator<char>(oss));
+
+  // https://unix.stackexchange.com/questions/631501
+  auto padLength = (4 - (oss.tellp() % 4)) % 4;
+  for (auto i = 0; i < padLength; ++i) {
```
Comment (aeulitz) [inline vnext/Common/Utilities.cpp:50]: Nit: Not sure if this is valid C++, but maybe you can avoid the for loop with something like

oss << ("\0===" + oss.tellp() % 4);

**Review by aeulitz — APPROVED** [review]

File: vnext/Common/Utilities.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+using std::string_view;
+using std::wstring_view;
+using winrt::array_view;
+
+using winrt::Windows::Security::Cryptography::BinaryStringEncoding;
+using winrt::Windows::Security::Cryptography::CryptographicBuffer;
+
+namespace Microsoft::React::Utilities {
+
+string DecodeBase64(string_view &&base64) noexcept {
```
[the remark this answers is not in the document]
Comment (JunielKatarn) [reply vnext/Common/Utilities.cpp:26]: True.
Will fix.

File: vnext/Common/Utilities.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  typedef array_view<char const> av_t;
+  auto bytes = av_t(text.data(), static_cast<av_t::size_type>(text.size()));
+
+  using namespace boost::archive::iterators;
+  typedef base64_from_binary<transform_width<const char *, 6, 8>> encode_base64;
+  std::ostringstream oss;
+  std::copy(encode_base64(bytes.cbegin()), encode_base64(bytes.cend()), ostream_iterator<char>(oss));
+
+  // https://unix.stackexchange.com/questions/631501
+  auto padLength = (4 - (oss.tellp() % 4)) % 4;
```
[the remark this answers is not in the document]
Comment (JunielKatarn) [reply vnext/Common/Utilities.cpp:49]: That's exactly the bug being fixed.

File: vnext/Common/Utilities.cpp (hunk 1 above)
Comment (JunielKatarn) [reply vnext/Common/Utilities.cpp:50]: Would that add the pad for cases where `tellp() % 4` = 1 and 2?

Comment (aeulitz) [reply vnext/Common/Utilities.cpp:50]: this suggestion was off the cuff & would need to be tested. the intention is that the offset into the string literal should mean that that "===" gets appended for tellp()%4==1, "==" for tellp()%4==2.

**Review by aeulitz — APPROVED** [review]

**Review by marlenecota — APPROVED** [review]

**Pull request merged**